tests/installation/select_patterns_and_packages.pm

```diff
@@ -155,113 +141,132 @@ sub package_action {
 
-            if (get_var('PATTERNS', '') =~ /default/ && !(get_var('PATTERNS', '') =~ /$current_pattern/)) {
-                $needs_to_be_selected = $selected;
-                record_info("keep default", "");
-            }
-            if (!$needs_to_be_selected && $selected) {
-                send_key ' ';
-                record_info("unselect", "");
-                assert_screen [qw(current-pattern-unselected current-pattern-autoselected)], 8;
+        if ($ret) {    # unneedled pattern
+            for my $p (keys %patterns) {
+                my $sel     = 1;
+                my $pattern = $p;    # store pattern untouched
+                if ($p =~ /^-/) {
+                    # this pattern shall be deselected as indicated by '-' prefix
+                    $sel = 0;
+                    $p =~ s/^-//;
+                }
+                if (match_has_tag("pattern-$p")) {
+                    $needs_to_be_selected = $sel;
+                    $current_pattern      = $p;
+                    # add pattern to the set of detected patterns
+                    $processed_patterns{$p} = undef;
+                    delete $patterns{$pattern};
+                    record_info($current_pattern, $needs_to_be_selected);
+                }
             }
-            movedownelseend;
-            check12qtbug;
+            last if ((get_var('PATTERNS', '') =~ /default/) && !(scalar keys %patterns));
         }
-        # check if we have processed all patterns mentioned in the test suite settings
-        # delete special 'all' and 'default' keys from the check
-        delete $patterns{default};
-        delete $patterns{all};
-        my @unseen;
-        foreach my $k (keys %patterns) {
-            #Remove - from the key in case of deselection
-            $k =~ s/^-//;
-            push @unseen, $k if (not exists $processed_patterns{$k});
+
+        $needs_to_be_selected = 1 if get_var('PATTERNS', '') =~ /all/;
+
+        my $selected = check_screen([qw(current-pattern-selected on-category)], 0);
+        if ($selected && $selected->{needle}->has_tag('on-category')) {
+            move_down;
+            next;
+        }
+        if ($needs_to_be_selected && !$selected) {
+            switch_selection(action => 'select', needles => ['current-pattern-selected']);
         }
+        if (get_var('PATTERNS', '') =~ /default/ && !(get_var('PATTERNS', '') =~ /$current_pattern/)) {
+            $needs_to_be_selected = $selected;
+            record_info("keep default", "");
+        }
+        if (!$needs_to_be_selected && $selected) {
+            switch_selection(action => 'unselect', needles => [qw(current-pattern-unselected current-pattern-autoselected)]);
+        }
+        move_down;
+        # check if we have processed all patterns mentioned in the test suite settings
+        my @unseen = keys %patterns;
         die "Not all patterns given in the job settings were processed:" . join(", ", @unseen) if @unseen;
```

Comment:
Looks much cleaner! I should have done same when worked on the ticket to fail the test when some patterns were not processed =)

Reply:
I've put this check outside `while` (I put inside by mistake) and I'm updating verification.